From ee87768254f12ce34e8d7ca53d874f5b67de26d7 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Wed, 5 Feb 2025 11:15:14 +0100
Subject: [PATCH] irqchip: stm32mp-exti1: correct IMR management

Correct IMR management to avoid side effect with secure world
if the associated BIT is not protected by RIF, no more read IMR register
but just update the bit managed by Linux driver in:
- stm32mp_exti_set_bit() and stm32mp_exti_clr_bit() to update mask_cache
- stm32mp_chip_suspend() and stm32mp_chip_resume() to
  manage the EXTI interruptions in IMR (unmask in RUN, wakeup in SUSPEND)
- remove the flag IRQCHIP_MASK_ON_SUSPEND, as wakeup source configuration
  facility is managed in EXTI driver with stm32mp_exti_set_wake()
  and chip_data->wake_active

This patch avoid issue with CPUIdle, based on PSCI OS initiated,
on STM32MP25 SoC: the active lp-timer 3 irq (not wake-up) are not
correctly restored after resume.

Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I3e88a46ab846331cb47da0c9c02c0736b2214caf
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/434652
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Jerome BESNARD <jerome.besnard@st.com>
---
 drivers/irqchip/irq-stm32mp-exti.c | 44 ++++++++++++++++++------------
 1 file changed, 27 insertions(+), 17 deletions(-)

--- a/drivers/irqchip/irq-stm32mp-exti.c
+++ b/drivers/irqchip/irq-stm32mp-exti.c
@@ -280,16 +280,21 @@ static int stm32mp_exti_convert_type(str
 }
 
 static void stm32mp_chip_suspend(struct stm32mp_exti_chip_data *chip_data,
-				 u32 wake_active)
+				 u32 mask_cache, u32 wake_active)
 {
 	const struct stm32mp_exti_bank *bank = chip_data->reg_bank;
 	void __iomem *base = chip_data->host_data->base;
+	u32 val;
 
 	/* save rtsr, ftsr registers */
 	chip_data->rtsr_cache = readl_relaxed(base + bank->rtsr_ofst);
 	chip_data->ftsr_cache = readl_relaxed(base + bank->ftsr_ofst);
 
-	writel_relaxed(wake_active, base + bank->imr_ofst);
+	/* mask active and set wakeup, preserve bits set by secure world */
+	val = readl_relaxed(base + bank->imr_ofst);
+	val &= ~mask_cache;
+	val |= wake_active;
+	writel_relaxed(val, base + bank->imr_ofst);
 
 	/* wakeup for IRQ power domain for S2IDLE */
 	if (wake_active)
@@ -297,16 +302,21 @@ static void stm32mp_chip_suspend(struct
 }
 
 static void stm32mp_chip_resume(struct stm32mp_exti_chip_data *chip_data,
-				u32 mask_cache)
+				u32 mask_cache, u32 wake_active)
 {
 	const struct stm32mp_exti_bank *bank = chip_data->reg_bank;
 	void __iomem *base = chip_data->host_data->base;
+	u32 val;
 
 	/* restore rtsr, ftsr, registers */
 	writel_relaxed(chip_data->rtsr_cache, base + bank->rtsr_ofst);
 	writel_relaxed(chip_data->ftsr_cache, base + bank->ftsr_ofst);
 
-	writel_relaxed(mask_cache, base + bank->imr_ofst);
+	/* mask wakeup and set active, preserve bits set by secure world */
+	val = readl_relaxed(base + bank->imr_ofst);
+	val &= ~wake_active;
+	val |= mask_cache;
+	writel_relaxed(val, base + bank->imr_ofst);
 
 	if (mask_cache)
 		pm_runtime_get(chip_data->host_data->dev);
@@ -326,26 +336,28 @@ static inline u32 stm32mp_exti_set_bit(s
 {
 	struct stm32mp_exti_chip_data *chip_data = irq_data_get_irq_chip_data(d);
 	void __iomem *base = chip_data->host_data->base;
-	u32 val;
+	u32 val, mask;
 
+	mask = BIT(d->hwirq % IRQS_PER_BANK);
 	val = readl_relaxed(base + reg);
-	val |= BIT(d->hwirq % IRQS_PER_BANK);
+	val |= mask;
 	writel_relaxed(val, base + reg);
 
-	return val;
+	return mask;
 }
 
 static inline u32 stm32mp_exti_clr_bit(struct irq_data *d, u32 reg)
 {
 	struct stm32mp_exti_chip_data *chip_data = irq_data_get_irq_chip_data(d);
 	void __iomem *base = chip_data->host_data->base;
-	u32 val;
+	u32 val, mask;
 
+	mask = BIT(d->hwirq % IRQS_PER_BANK);
 	val = readl_relaxed(base + reg);
-	val &= ~BIT(d->hwirq % IRQS_PER_BANK);
+	val &= ~mask;
 	writel_relaxed(val, base + reg);
 
-	return val;
+	return mask;
 }
 
 static void stm32mp_exti_eoi(struct irq_data *d)
@@ -362,7 +374,7 @@ static void stm32mp_exti_eoi(struct irq_
 	if (!chip_data->mask_cache)
 		pm_runtime_get(chip_data->host_data->dev);
 
-	chip_data->mask_cache = stm32mp_exti_set_bit(d, bank->imr_ofst);
+	chip_data->mask_cache |= stm32mp_exti_set_bit(d, bank->imr_ofst);
 
 	raw_spin_unlock(&chip_data->rlock);
 
@@ -375,7 +387,7 @@ static void stm32mp_exti_mask(struct irq
 	const struct stm32mp_exti_bank *bank = chip_data->reg_bank;
 
 	raw_spin_lock(&chip_data->rlock);
-	chip_data->mask_cache = stm32mp_exti_clr_bit(d, bank->imr_ofst);
+	chip_data->mask_cache &= ~stm32mp_exti_clr_bit(d, bank->imr_ofst);
 
 	/* power domain is OFF when IMR becomes 0 */
 	if (!chip_data->mask_cache)
@@ -397,7 +409,7 @@ static void stm32mp_exti_unmask(struct i
 	if (!chip_data->mask_cache)
 		pm_runtime_get(chip_data->host_data->dev);
 
-	chip_data->mask_cache = stm32mp_exti_set_bit(d, bank->imr_ofst);
+	chip_data->mask_cache |= stm32mp_exti_set_bit(d, bank->imr_ofst);
 	raw_spin_unlock(&chip_data->rlock);
 
 	irq_chip_unmask_parent(d);
@@ -470,7 +482,7 @@ static int stm32mp_exti_suspend(struct d
 
 	for (i = 0; i < host_data->drv_data->bank_nr; i++) {
 		chip_data = &host_data->chips_data[i];
-		stm32mp_chip_suspend(chip_data, chip_data->wake_active);
+		stm32mp_chip_suspend(chip_data, chip_data->mask_cache, chip_data->wake_active);
 	}
 
 	return 0;
@@ -505,7 +517,7 @@ static int stm32mp_exti_resume(struct de
 	stm32mp_exti_resume_gpio_mux(host_data);
 	for (i = 0; i < host_data->drv_data->bank_nr; i++) {
 		chip_data = &host_data->chips_data[i];
-		stm32mp_chip_resume(chip_data, chip_data->mask_cache);
+		stm32mp_chip_resume(chip_data, chip_data->mask_cache, chip_data->wake_active);
 	}
 
 	return 0;
@@ -534,7 +546,6 @@ static struct irq_chip stm32mp_exti_chip
 	.irq_retrigger		= stm32mp_exti_retrigger,
 	.irq_set_type		= stm32mp_exti_set_type,
 	.irq_set_wake		= stm32mp_exti_set_wake,
-	.flags			= IRQCHIP_MASK_ON_SUSPEND,
 	.irq_set_affinity	= IS_ENABLED(CONFIG_SMP) ? irq_chip_set_affinity_parent : NULL,
 };
 
@@ -549,7 +560,6 @@ static struct irq_chip stm32mp_exti_chip
 	.irq_retrigger		= irq_chip_retrigger_hierarchy,
 	.irq_set_type		= irq_chip_set_type_parent,
 	.irq_set_wake		= stm32mp_exti_set_wake,
-	.flags			= IRQCHIP_MASK_ON_SUSPEND,
 	.irq_set_affinity	= IS_ENABLED(CONFIG_SMP) ? irq_chip_set_affinity_parent : NULL,
 };
 
